src/decisionengine/framework/taskmanager/TaskManager.py

```diff
@@ -180,74 +207,94 @@ def wait_for_all(self, events_done):
-                self.take_offline()
 
-        self.logger.info(f"Task Manager {self.id} received a stop signal and is exiting")
+        _cl = Connection(self.broker_url)
+        with connections[_cl].acquire(block=True) as conn:
+
+            def run_cycle(body, message):
+                module_spec = body["source_module"]
+                module_name = body["class_name"]
+                data = body["data"]
+                self.logger.debug(f"Data received from {module_name}: {data}")
+                if not data:
+                    self.logger.warning("No data acquired")
+                    message.ack()
+                    return
+
+                if not self.sources_have_run_once:
+                    self.source_product_cache.update(**data)
+                    missing_products = self.expected_source_products - set(self.source_product_cache.keys())
+                    if missing_products:
+                        self.logger.info(f"Waiting on more data (missing {missing_products})")
+                        message.ack()
+                        return
+                    data = self.source_product_cache
+                    self.sources_have_run_once = True
+
+                header = datablock.Header(
+                    self.data_block_t0.taskmanager_id, create_time=time.time(), creator=module_spec
+                )
+                self.logger.info(f"Source {module_name} header done")
+                self.data_block_put(data, header, self.data_block_t0)
+                self.logger.info(f"Source {module_name} data block put done")
+
+                self.logger.info(f"Message on bus: {body}")
+                try:
+                    self.decision_cycle()
+                    with self.state.lock:
+                        if not self.state.should_stop():
+                            # If we are signaled to stop, don't override that state
+                            # otherwise the last decision_cycle completed without error
+                            self.state.set(State.STEADY)
+                except Exception:  # pragma: no cover
+                    self.logger.exception("Exception in the task manager main loop")
+                    self.logger.error(
+                        "Error occured. Task manager %s exits with state %s", self.id, self.get_state_name()
+                    )
+                message.ack()
+
+            with conn.Consumer(self.queues.values(), accept=["pickle"], callbacks=[run_cycle]):
+                while not self.state.should_stop():
+                    try:
+                        # If source has been brought offline while the drain_events method is
+                        # executing, it will stall.  We therefore impose an arbitrary 5-second
+                        # timeout so that the 'should_stop()' method called in the while condition
+                        # will yield false and thus terminate the loop.
+                        conn.drain_events(timeout=5)
```

Comment:
Can I get a logger.debug mentioning where it is listening for events?

Reply:
Will do.